File: frontend/apps/ios/Pantopus/Core/Networking/Models/Membership/MembershipDTOs.swift
//
//  MembershipDTOs.swift
//  Pantopus
//
//  Decoder shapes for `GET /api/personas/:id/membership` — the fan-side
//  view of their own membership (A10.8). Built by `serializeMembershipForFan`
//  (`backend/serializers/identitySerializers.js:352`), wrapped in
//  `stripNullish`, so several keys (e.g. `fanHandle`, period bounds) may be
//  absent — every field is therefore optional. The embedded `persona` is the
//  shared `serializeAudienceProfileForViewer` shape (camelCase keys).
//

import Foundation

// MARK: - GET /api/personas/:id/membership  (+ POST .../membership/cancel)

/// Envelope for the membership read and the cancel mutation (the cancel
/// route echoes the re-fetched membership in the same shape).
public struct PersonaMembershipResponse: Decodable, Sendable {
    public let membership: PersonaMembershipDTO?
}

public struct PersonaMembershipDTO: Decodable, Sendable {
    public let membershipId: String?
    public let persona: MembershipPersonaDTO?
    public let tier: MembershipTierDTO?
    public let status: String?
    public let cancelAtPeriodEnd: Bool?
    public let currentPeriodStart: String?
    public let currentPeriodEnd: String?
    // `scheduledTierChange` + `quotaRemaining` are present on the wire but
    // unused by the fan manage surface — Codable ignores the extra keys.
}

/// The persona the fan supports — `serializeAudienceProfileForViewer`
/// (camelCase). Only the fields the membership card binds to are decoded.
public struct MembershipPersonaDTO: Decodable, Sendable {
    public let id: String?
    public let handle: String?
    public let displayName: String?
    public let avatarUrl: String?
    public let category: String?
    public let audienceLabel: String?
    public let followerCount: Int?
    public let credential: CredentialDTO?

    public struct CredentialDTO: Decodable, Sendable {
        public let status: String?
        public let label: String?
    }
}

/// The fan's tier. Perk fields (`msgThreadsPerPeriod`, `creatorCanInitiateDm`,
/// `replyPolicy`) drive the "What you get" benefit rows. Backend emits
/// camelCase here.
public struct MembershipTierDTO: Decodable, Sendable {
    public let id: String?
    public let rank: Int?
    public let name: String?
    public let priceCents: Int?
    public let currency: String?
    public let billingInterval: String?
    public let msgThreadsPerPeriod: Int?
    public let creatorCanInitiateDm: Bool?
    public let replyPolicy: String?
}

File: frontend/apps/android/app/src/main/java/app/pantopus/android/data/api/models/membership/MembershipDtos.kt
@file:Suppress("PackageNaming")

package app.pantopus.android.data.api.models.membership

import com.squareup.moshi.JsonClass

/**
 * Decoder shapes for `GET /api/personas/:id/membership` — the fan-side view
 * of their own membership (A10.8). Built by `serializeMembershipForFan`
 * (`backend/serializers/identitySerializers.js:352`) wrapped in `stripNullish`,
 * so several keys may be absent — every field is optional. The membership
 * serializer emits camelCase keys, so no `@Json` mapping is needed. The cancel
 * route echoes the same envelope.
 */

@JsonClass(generateAdapter = true)
data class PersonaMembershipResponse(
    val membership: PersonaMembershipDto? = null,
)

@JsonClass(generateAdapter = true)
data class PersonaMembershipDto(
    val membershipId: String? = null,
    val persona: MembershipPersonaDto? = null,
    val tier: MembershipTierDto? = null,
    val status: String? = null,
    val cancelAtPeriodEnd: Boolean? = null,
    val currentPeriodStart: String? = null,
    val currentPeriodEnd: String? = null,
)

/** The persona the fan supports — shared `serializeAudienceProfileForViewer`. */
@JsonClass(generateAdapter = true)
data class MembershipPersonaDto(
    val id: String? = null,
    val handle: String? = null,
    val displayName: String? = null,
    val avatarUrl: String? = null,
    val category: String? = null,
    val audienceLabel: String? = null,
    val followerCount: Int? = null,
    val credential: CredentialDto? = null,
)

@JsonClass(generateAdapter = true)
data class CredentialDto(
    val status: String? = null,
    val label: String? = null,
)

/** The fan's tier — perk fields drive the "What you get" benefit rows. */
@JsonClass(generateAdapter = true)
data class MembershipTierDto(
    val id: String? = null,
    val rank: Int? = null,
    val name: String? = null,
    val priceCents: Int? = null,
    val currency: String? = null,
    val billingInterval: String? = null,
    val msgThreadsPerPeriod: Int? = null,
    val creatorCanInitiateDm: Boolean? = null,
    val replyPolicy: String? = null,
)
